File: 剑指offer/54.二叉搜索树的第k大节点.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def kthLargest(self, root: TreeNode, k: int) -> int:
        ans = None

        def help(root):
            nonlocal k, ans
            if ans or not root:
                return

            help(root.right)
            if not ans and k == 1:
                ans = root.val
            else:
                k -= 1
            help(root.left)

        help(root)
        return ans

File: 剑指offer/54.二叉搜索树的第k大节点.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    int ans = 0;
    int k = 0;

    void help(TreeNode* root)
    {
        if (root == NULL or k == 0) {
            return;
        }

        help(root->right);

        k -= 1;
        if (k == 0) {
            ans = root->val;
        }

        help(root->left);
    }

    int kthLargest(TreeNode* root, int k)
    {
        this->k = k;
        help(root);
        return ans;
    }
};